sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/StartingPointShardsFinder.java

```diff
@@ -0,0 +1,172 @@
+   *                             /           \
+   *                  0005------+             0009------+
+   *                 /                                   \
+   * 0002-----------+                                     0010------
+   *                 \                                   /
+   *                  0006------------------------------+
+   * </pre>
+   */
+  Iterable<Shard> findShardsAtStartingPoint(SimplifiedKinesisClient kinesis, String streamName,
+      StartingPoint startingPoint) throws TransientKinesisException {
+    List<Shard> allShards = kinesis.listShards(streamName);
+    Set<Shard> initialShards = findInitialShardsWithoutParents(streamName, allShards);
+
+    Set<Shard> startingPointShards = new HashSet<>();
+    Sets.SetView<Shard> expiredShards;
+    do {
+      Set<Shard> validShards = validateShards(kinesis, initialShards, streamName, startingPoint);
+      startingPointShards.addAll(validShards);
+      expiredShards = Sets.difference(initialShards, validShards);
+      Set<String> expiredShardIds = new HashSet<>();
+      for (Shard expiredShard : expiredShards) {
+        expiredShardIds.add(expiredShard.getShardId());
+      }
+      if (!expiredShardIds.isEmpty()) {
+        LOGGER.info("Following shards expired for {} stream at '{}' starting point: {}", streamName,
+            startingPoint, expiredShardIds);
+      }
+      initialShards = new HashSet<>();
+      for (Shard shard : allShards) {
+        if (expiredShardIds.contains(shard.getParentShardId())) {
+          initialShards.add(shard);
+        }
+      }
+    } while (!expiredShards.isEmpty());
+    return startingPointShards;
+  }
+
+  /**
+   * Finds the initial set of shards (the oldest ones). These shards do not have their parents in
+   * the shard list.
+   */
+  private Set<Shard> findInitialShardsWithoutParents(String streamName, List<Shard> allShards) {
+    Set<String> shardIds = new HashSet<>();
+    for (Shard shard : allShards) {
+      shardIds.add(shard.getShardId());
+    }
+    LOGGER.info("Stream {} has following shards: {}", streamName, shardIds);
+    Set<Shard> shardsWithoutParents = new HashSet<>();
+    for (Shard shard : allShards) {
+      if (!shardIds.contains(shard.getParentShardId())) {
+        shardsWithoutParents.add(shard);
+      }
+    }
+    return shardsWithoutParents;
+  }
+
+  private Set<Shard> validateShards(SimplifiedKinesisClient kinesis, Iterable<Shard> rootShards,
+      String streamName, StartingPoint startingPoint)
+      throws TransientKinesisException {
+    Set<Shard> validShards = new HashSet<>();
+    ShardIteratorType shardIteratorType = ShardIteratorType
+        .fromValue(startingPoint.getPositionName());
+    for (Shard shard : rootShards) {
+      String shardIterator = kinesis.getShardIterator(streamName, shard.getShardId(),
+          shardIteratorType, null, startingPoint.getTimestamp());
+      GetKinesisRecordsResult records = kinesis.getRecords(shardIterator, streamName,
```

Comment:
This does not preclude these records from being later returned via the reader, because we're not acking them, correct? (or is it because we're starting from a particular timestamp?) Could you add a comment?

Reply:
The records can be read again and again from the same point when in the retention period. There's no concept of acking records in Kinesis and we also don't do any Beam acking here.